kernel/comps/framebuffer/src/framebuffer.rs

```diff
@@ -0,0 +1,219 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::sync::Arc;
+
+use ostd::{boot::boot_info, io::IoMem, mm::VmIo};
+use spin::Once;
+
+/// The framebuffer used for text or graphical output.
+///
+/// # Notes
+///
+/// It is highly recommended to use a synchronization primitive, such as a `SpinLock`, to
+/// lock the framebuffer before performing any operation on it.
+/// Failing to properly synchronize access can result in corrupted framebuffer content
+/// or undefined behavior during rendering.
+#[derive(Debug)]
+pub struct FrameBuffer {
+    io_mem: IoMem,
+    width: usize,
+    height: usize,
+    pixel_format: PixelFormat,
+}
+
+/// Pixel format that defines the memory layout of each pixel in the framebuffer.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum PixelFormat {
+    /// Each pixel uses 8 bits to represent its grayscale intensity, ranging from 0 (black) to 255 (white).
+    Grayscale8,
+    /// Each pixel uses 16 bits, with 5 bits for Red, 6 bits for Green, and 5 bits for Blue.
+    Rgb565,
+    /// Each pixel uses 24 bits, with 8 bits for Red, 8 bits for Green, and 8 bits for Blue.
+    Rgb888,
+    /// Each pixel uses 32 bits, with 8 bits each for Red, Green, Blue, and Alpha (transparency).
+    Rgba,
+    /// Each pixel uses 32 bits, with 8 bits each for Green, Blue, Red, and Alpha (transparency).
+    Gbra,
+}
+
+/// Individual pixel data containing raw channel values.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct Pixel {
+    pub red: u8,
+    pub green: u8,
+    pub blue: u8,
+    pub alpha: u8,
+}
+
+pub static FRAMEBUFFER: Once<Arc<FrameBuffer>> = Once::new();
+
+pub(crate) fn init() {
+    let Some(framebuffer_arg) = boot_info().framebuffer_arg else {
+        log::warn!("Framebuffer not found");
+        return;
+    };
+
+    if framebuffer_arg.address == 0 {
+        log::error!("Framebuffer address is zero");
+        return;
+    }
+
+    let pixel_format = match framebuffer_arg.bpp {
+        8 => PixelFormat::Grayscale8,
+        16 => PixelFormat::Rgb565,
+        24 => PixelFormat::Rgb888,
+        32 => {
+            // FIXME: There are several pixel formats that have the same BPP. We lost the information
+            // during the boot phase, so here we guess the pixel format on a best effort basis.
+            PixelFormat::Rgba
+        }
+        _ => {
+            log::error!(
+                "Unsupported framebuffer pixel format: {} bpp",
+                framebuffer_arg.bpp
+            );
+            return;
+        }
+    };
+
+    let framebuffer = {
+        let fb_base = framebuffer_arg.address;
+        let fb_size = framebuffer_arg.width * framebuffer_arg.height * (framebuffer_arg.bpp / 8);
```

Comment:
```suggestion
        let fb_size = framebuffer_arg.width * framebuffer_arg.height * (framebuffer_arg.bpp / u8::BITS);
```

It seems that `8` can represent many things. It can represent bits per byte, the font width, and the font height. I suggest using `u8::BITS` for the first and defining constants for the last two:
```rust
const FONT_WIDTH: usize = 8;
const FONT_HEIGHT: usize = 8;
```